contracts/FungibleToken-v2.cdc

```diff
@@ -0,0 +1,197 @@
+/**
+
+# The Flow Fungible Token standard
+
+## `FungibleToken` contract
+
+The Fungible Token standard is no longer an interface
+that all fungible token contracts would have to conform to.
+
+If a users wants to deploy a new token contract, their contract
+does not need to implement the FungibleToken interface, but their tokens
+do need to implement the interfaces defined in this contract.
+
+## `Vault` resource interface
+
+Each fungible token resource type needs to implement the `Vault` resource interface.
+
+## `Provider`, `Receiver`, and `Balance` resource interfaces
+
+These interfaces declare pre-conditions and post-conditions that restrict
+the execution of the functions in the Vault.
+
+They are separate because it gives the user the ability to share
+a reference to their Vault that only exposes the fields functions
+in one or more of the interfaces.
+
+It also gives users the ability to make custom resources that implement
+these interfaces to do various things with the tokens.
+For example, a faucet can be implemented by conforming
+to the Provider interface.
+
+*/
+
+/// FungibleToken
+///
+pub contract FungibleToken {
+
+    /// Provider
+    ///
+    /// The interface that enforces the requirements for withdrawing
+    /// tokens from the implementing type.
+    ///
+    /// It does not enforce requirements on `balance` here,
+    /// because it leaves open the possibility of creating custom providers
+    /// that do not necessarily need their own balance.
+    ///
+    pub resource interface Provider {
```

Comment:
Not sure how we'd make it happen, but it would be nice if providers had some kind of way to be scoped to a specific amount like an allowance. Curious if there's any thought to that in the V2 standards since that's a pretty prominent risk that lots of dapps have out there right now.

An attempt to wrap providers to accomplish this is in a PR here (needs lots of tweaking still, though):
https://github.com/Flowtyio/flow-utils/pull/6

Reply:
yeah, I'd love to have this too. I was holding off on this to focus on the core part of the proposal, but maybe I'll add it to this anyway

Reply:
I'll copy yours over and make some edits

Reply:
@bluesign had some feedback on this (one big one is making it a resource instead of a struct) and might want to weigh in